### ISSUE TITLE
3D model rotation and zoom does not smooth?

### ISSUE THREAD

**Issue by tuzenvn** [viromedia/virocore#160]
I try to build ARRetail with my model like lamps and chairs but I see interactions with models (rotation and zoom) does not smooth like IOS ARToolkit. My issue sometimes conflig actions, for example when I using two fingers to rotation but model do zoom out or zoom in and when I want to scale model then that do rotation model.

**Comment by radvani** [comment]
Hi @tuzenvn,

Could you let me know what device you're using for the ARRetail example? I'd like to see if I can reproduce this. Thanks.

**Comment by tuzenvn** [comment]
Hi @radvani ,

I use Pixel XL - OS 8.1 for testing.

Thanks.

**Comment by HLneoh** [comment]
@achuvm What if I want to have smooth 3D scaling and rotating in my AR application? Is it possible to achieve that?

**Comment by achuvm** [comment]
@HLneoh,

Check out the code for two-finger scaling/rotation in our open source Figment App:
https://github.com/viromedia/figment-ar

You can install the app and test it for yourself from the Play Store or App Store.

**Comment by HLneoh** [comment]
@achuvm Thanks for the comment. However, figment AR used viro react and react native to implement the feature. May I know is it possible to use viro core (Java) to achieve that?

**Comment by achuvm** [comment]
Hi @HLneoh,

Yes, you can use very similar logic (unfortunately we don't have a published example of this) through the use of the [GesturePinchListener](https://developer.viromedia.com/virocore/reference/com/viro/core/GesturePinchListener.html) and the [GestureRotateListener](https://developer.viromedia.com/virocore/reference/com/viro/core/GestureRotateListener.html).

Let us know if you cant get it to work!

**Comment by HLneoh** [comment]
@achuvm Yes, I am able to achieve that. However, I cannot get a smooth rotating and scaling. My issue sometimes conflig actions, for example when I using two fingers to rotation but model do zoom out or zoom in and when I want to scale model then that do rotation model.

**Comment by achuvm** [comment]
Hi @HLneoh,

Hmm, that sounds like an issue with the callbacks if the rotating and pinch gestures are being mis-characterized by the callbacks. We'll have to take a look at this and fix any issues in a future release.

In the meantime, you can add your own `OnTouchListener` by calling [`setOnTouchListener()`](https://developer.android.com/reference/android/view/View.html#setOnTouchListener(android.view.View.OnTouchListener)) on `ViroViewARCore`. That way you can use any public code out there to process Android touch events. These will compute 2D coordinates so you may need a way for the user to tell you which object should be rotated/scaled.

Thanks,

**Comment by dthian** [comment]
Hey @HLneoh, as Andy mentioned, you can add your own OnTouchListener to handle the rotate/gesture listeners yourself (we actually use Android's built in rotate/gesture listeners to trigger the gesture callback events). 

Else, you should also have all the information you need (pinch / rotation values) to determine your own custom threshold, and thus to toggle between rotating or scaling your 3D object.

**Comment by lezan** [comment]
> 
> 
> @achuvm Yes, I am able to achieve that. However, I cannot get a smooth rotating and scaling. My issue sometimes conflig actions, for example when I using two fingers to rotation but model do zoom out or zoom in and when I want to scale model then that do rotation model.

Is it still an issue? Because I am experience the same problem.
